pkg/controller/registry/resolver/solver/dict.go

```diff
@@ -191,36 +184,77 @@ func compileDict(installables []Installable) *dict {
 		}
 	}
 
-	d.cards = d.c.CardSort(weights)
-
 	return &d
 }
 
-// Solve searches for a solution to the problem encoded in a dict that
-// minimizes the total weight of all selected Installables.
-func (d *dict) Solve(ctx context.Context, g *gini.Gini) int {
+func (d *dict) ApplyConstraints(g *gini.Gini) {
 	d.c.ToCnf(g)
+	for m := range d.constraints {
+		g.Assume(m)
+	}
+}
 
-	var result int
-	linearSearch(0, d.cards.N(), func(w int) bool {
-		if w >= 0 {
-			// Omitting the cardinality constraint on the
-			// last iteration prevents the cardinality
-			// constraint literal from showing up in the
-			// set of failed assumptions should there be
-			// no solution.
-			g.Assume(d.cards.Leq(w))
+type litSet interface {
+	Contains(z.Lit) bool
+}
+
+// Given an inter.S for which the last call to Test returned 1 and a
+// set of lits to fix, attempts to minimize the cardinality of the
+// solution without removing any of the fixed lits.
+func (d *dict) MinimizeCardinality(g inter.S, assumed litSet) error {
+	var extras []z.Lit
+	for m := range d.installables {
+		if g.Value(m) && !assumed.Contains(m) {
+			extras = append(extras, m)
+		}
+	}
+	cards := d.c.CardSort(extras)
```

Comment:
I think there needs to be a call to https://pkg.go.dev/github.com/irifrance/gini/logic?tab=doc#C.CnfSince to add any new clauses. Looking into it...